ui/prometheus-plugin/src/plugins/PrometheusDatasourceEditor/PrometheusDatasourceEditor.tsx

```diff
@@ -0,0 +1,327 @@
+          />
+          <Typography py={1} variant="h4">
+            Allowed endpoints
+          </Typography>
+          {value.proxy?.spec.allowed_endpoints &&
+            value.proxy.spec.allowed_endpoints.map(({ endpoint_pattern, method }, i) => {
+              return (
+                <Grid container key={i} spacing={2} mb={2}>
+                  <Grid item xs={4}>
+                    <TextField
+                      disabled // at the moment the allowed endpoints cannot be modified (enforced by backend)
+                      fullWidth
+                      label="Endpoint pattern"
+                      value={endpoint_pattern}
+                      InputProps={{
+                        readOnly: isReadonly,
+                      }}
+                    />
+                  </Grid>
+                  <Grid item xs={4}>
+                    <TextField
+                      disabled // at the moment the allowed endpoints cannot be modified (enforced by backend)
+                      fullWidth
+                      label="URL"
+                      value={method}
+                      InputProps={{
+                        readOnly: isReadonly,
+                      }}
+                    />
+                  </Grid>
+                </Grid>
+              );
+            })}
+          <Typography py={1} variant="h4">
+            Request Headers
+          </Typography>
+          {value.proxy?.spec.headers !== undefined &&
+            Object.keys(value.proxy.spec.headers).map((headerName, i) => {
+              return (
+                <Grid container key={i} spacing={2} mb={2}>
+                  <Grid item xs={4}>
+                    <TextField
+                      fullWidth
+                      label="Header name"
+                      value={headerName}
+                      InputProps={{
+                        readOnly: isReadonly,
+                      }}
+                      onChange={(e) =>
+                        onChange({
+                          ...value,
+                          ...(value.proxy && {
+                            proxy: {
+                              ...value.proxy,
+                              spec: {
+                                ...value.proxy.spec,
+                                headers: buildNewHeaders(value.proxy.spec.headers, headerName, e.target.value),
+                              },
+                            },
+                          }),
+                        })
+                      }
+                    />
+                  </Grid>
+                  <Grid item xs={4}>
+                    <TextField
+                      fullWidth
+                      label="Header value"
+                      value={value.proxy?.spec.headers?.[headerName]}
+                      InputProps={{
+                        readOnly: isReadonly,
+                      }}
+                      onChange={(e) =>
+                        onChange({
```

Comment:
Could try immer in these forms too to avoid all the spread operator usage, see `useImmer` in VariableEditForm as example

Reply:
at first I used it but the performance where quite bad, maybe not because of useImmer specifically though since I changed other things since. I will try again to use it & see.

Reply:
Yeah worth a try, but depending what comes out of [this conversation](https://github.com/perses/perses/pull/1282#discussion_r1268626997) about form / validation libraries we could wait and refactor later I guess

Reply:
Yes if you don't mind I'd like to tackle this one in a follow-up PR.